rcl/include/rcl/publisher.h

```diff
@@ -0,0 +1,266 @@
+ *
+ * TODO(wjwwood): update after researching what the blocking behavior is
+ *                or optionally link to a document that describes blocking
+ *                behavior is more detail.
+ * Calling rcl_publish is a potentially blocking call.
+ * When called rcl_publish will immediately do any publishing related work,
+ * including, but not limited to, converting the message into a different type,
+ * serializing the message, collecting publish statistics, etc.
+ * The last thing it will do is call the underlying middleware's publish
+ * function which may or may not block based on the quality of service settings
+ * given via the publisher options in rcl_publisher_init().
+ * For example, if the reliability is set to reliable, then a publish may block
+ * until space in the publish queue is available, but if the reliability is set
+ * to best effort then it should not block.
+ *
+ * The ROS message given by the ros_message void pointer is always owned by the
+ * calling code, but should remain constant during publish.
+ *
+ * This function is thread safe so long as access to both the publisher and the
+ * ros_message is synchronized.
+ * That means that calling rcl_publish from multiple threads is allowed, but
+ * calling rcl_publish at the same time as non-thread safe publisher functions
+ * is not, e.g. calling rcl_publish and rcl_publisher_fini concurrently
+ * is not allowed.
+ * Before calling rcl_publish the message can change and after calling
+ * rcl_publish the message can change, but it cannot be changed during the
+ * publish call.
+ * The same ros_message, however, can be passed to multiple calls of
+ * rcl_publish simultaneously, even if the publishers differ.
+ * The ros_message is unmodified by rcl_publish.
+ *
+ * \param[in] publisher handle to the publisher which will do the publishing
+ * \param[in] ros_message type-erased pointer to the ROS message
+ * \return RCL_RET_OK if the message was published successfully, or
+ *         RCL_RET_INVALID_ARGUMENT if any arugments are invalid, or
+ *         RCL_RET_PUBLISHER_INVALID if the publisher is invalid, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+rcl_ret_t
+rcl_publish(const rcl_publisher_t * publisher, const void * ros_message);
+
+/// Get the topic name for the publisher.
+/* This function returns the publisher's internal topic name string.
+ * This function can fail, and therefore return NULL, if the:
+ *   - publisher is NULL
+ *   - publisher is invalid (never called init, called fini, or invalid node)
+ *
+ * The returned string is only valid as long as the rcl_publisher_t is valid.
+ * The value of the string may change if the topic name changes, and therefore
+ * copying the string is recommended if this is a concern.
+ *
+ * This function is not thread-safe, and copying the result is not thread-safe.
+ * \TODO(wjwwood): should we provide a thread-safe copy_topic_name?
```

Comment:
On the topic of thread safety, I would say that wrapping this in synchronization is the responsibility of the language client library, because of how synchronization varies between libraries.